notion-distcopy/src/main/scala/com/ambiata/notion/distcopy/FileDistCopy.scala

```diff
@@ -0,0 +1,92 @@
+package com.ambiata.notion.distcopy
+
+import com.ambiata.mundane.control.RIO
+import com.ambiata.mundane.io._
+import com.ambiata.notion.core._
+import com.ambiata.poacher.hdfs.Hdfs
+import com.ambiata.saws.s3.{S3Address, S3Prefix, _}
+import org.apache.hadoop.fs._
+
+import scalaz.Scalaz._
+
+/**
+ * Copy functions between Hdfs and S3
+ */
+object FileDistCopy {
+
+  /**
+   * Download a large file by doing a distcopy from s3 to hdfs
+   * Don't do anything if the file already exists
+   */
+  def distCopyDownloadFile(from: S3Address, to: Location, locationIO: LocationIO): RIO[Unit] =
+    for {
+      toUri  <- hadoopPath(to)
+      exists <- pathExist(new Path(toUri), locationIO)
+      _      <- DistCopyJob.run(Mappings(Vector(DownloadMapping(from, new Path(toUri)))), distCopyConfiguration(locationIO)).whenM(exists)
+    } yield ()
+
+  /**
+   * Download a large directory by doing a distcopy from s3 to a hdfs directory
+   *
+   * Don't copy files which have already been copied
+   */
+  def distCopyDownloadDirectory(from: S3Prefix, to: Location, locationIO: LocationIO): RIO[Unit] =
+    for {
+      toUri         <- hadoopPath(to)
+      fromAddresses <- from.listAddress.execute(locationIO.s3Client)
+      mappings      <- fromAddresses.traverseU(createDownloadMapping(HdfsLocation(toUri), locationIO))
+      _             <- DistCopyJob.run(Mappings(mappings.toVector.flatten), distCopyConfiguration(locationIO))
+    } yield ()
+
+  /** upload a large file by doing a distcopy from hdfs to s3 */
+  def distCopyUploadFile(from: Location, to: S3Pattern, locationIO: LocationIO): RIO[Unit] =
+    for {
+      fromUri <- hadoopPath(from)
+      _       <- DistCopyJob.run(Mappings(Vector(UploadMapping(new Path(fromUri), S3Address(to.bucket, to.unknown)))), distCopyConfiguration(locationIO))
+    } yield ()
+
+  /** upload a large file by doing a distcopy from hdfs to s3 */
+  def distCopyUploadDirectory(from: Location, to: S3Location, locationIO: LocationIO): RIO[Unit] =
+    for {
+      fromUri   <- hadoopPath(from)
+      fromPaths <- Hdfs.globFiles(new Path(fromUri, "*")).run(locationIO.configuration)
+      mappings  <- fromPaths.traverseU(createUploadMapping(to, locationIO))
+      _         <- DistCopyJob.run(Mappings(mappings.toVector.flatten), distCopyConfiguration(locationIO))
+    } yield ()
```

Comment:
I think we should probably discuss how these are implemented offline - I don't think the types align with the implementation - as in there is assumptions about what type of Location - and I am really concerned that someone could use it and end up with more path related bugs. I could be missing something though - so can we discuss quickly at some point tomorrow the intentions etc...